### PR TITLE
Validate block settings for edit() and save()

### LINKED ISSUE #362: Validate usable render behavior on registered block
When registering a block, we should ensure that it has at minimum a `save` implementation (front-end UI representation) and optionally `edit`.

Once validated, we can collapse logic here to return early only on `settings` being falsey and assume if settings exist that a `edit` or `save` function/component can be used for displaying the block:

https://github.com/WordPress/gutenberg/blob/15db036/editor/editor/mode/visual.js#L25-L32

### PR DESCRIPTION
**Pull request by BE-Webdesign**
Fixes #362. Validates that a block has a `save()` and `edit()`.  This is
probably a temporary solution.  The API should probably be cleaned up a
bit so everything is not quite as coupled and tests can be done
differently.

### CODE REVIEW

File: blocks/api/registration.js [hunk 1]
```diff
@@ -104,3 +111,34 @@ export function getBlockSettings( slug ) {
...
+		);
+		return false;
+	}
+
+	if ( ! settings.edit || 'function' !== typeof settings.edit ) {
+		console.error(
+			'Block settings must specify a edit() method for each block.'
+		);
+		return false;
+	}
```
Comment (youknowriad) [inline blocks/api/registration.js]: Since we're adding some validation, should we validate all the mandatory attributes we have? I'm thinking of `icon`, `title` and `category`. Maybe we should also validate that the category is a known category.

Comment (BE-Webdesign) [inline blocks/api/registration.js]: I think this is fine for now, this sets our interface as requiring `save()` and `edit()`.

If there is buy-in to do more thorough validation, then I would probably redo this entire approach to become a bit more programmatic.  The more and more validation we add will make things more and more sprawly, so we will probably eventually want to introduce a schema.  A schema could double serve to help validate incoming attributes for an actual block as well.

File: blocks/api/registration.js [hunk 2]
```diff
@@ -43,6 +43,13 @@ export function registerBlock( slug, settings ) {
 		);
 		return;
 	}
+	if ( true !== validateBlockSettings( settings ) ) {
```
Comment (aduth) [inline blocks/api/registration.js]: There was a similar discussion in the original validation introduced about how to name this function based on its behavior. Eventually it appears we inlined the validation. A few questions then:

- Should we inline block settings validation for consistency with the slug validation?
- Should we expect the `validate` to _do_ something, like `throw`, or perhaps prefixing with `is` to reflect its boolean return value is more accurate, like `isValidBlockSettings`
- If we know the function returns a boolean, do we need to be so explicit with the comparison, or can we simplify to `! isValidBlockSettings` ?

File: blocks/api/registration.js [hunk 3]
```diff
@@ -43,6 +43,13 @@ export function registerBlock( slug, settings ) {
...
+		console.error(
+			'Block not registered.'
+		);
+		// Return the block even though it was not registered.
+		return Object.assign( { slug }, settings );
```
Comment (aduth) [inline blocks/api/registration.js]: Thinking since we already assign this as `block` below that we could move the condition after the variable assignment and simply return the variable instead of duplicating the logic (especially in case we change the merge logic down the road).

```js
const block = Object.assign( { slug }, settings );
if ( ! isValidBlockSettings( settings ) ) {
	console.error(
		'Block not registered.'
	);
	// Return the block even though it was not registered.
	return block;
}
```

Or perhaps we reflect the intent that the global `blocks` object should only assign the new block if it's valid, but allow the function to continue to the final `return`:

```js
const block = Object.assign( { slug }, settings );
if ( isValidBlockSettings( settings ) ) {
	blocks[ slug ] = block;
} else {
	console.error(
		'Block not registered.'
	);
}

return block;
```

File: blocks/api/test/function-refs.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/**
```
Comment (aduth) [inline blocks/api/test/function-refs.js]: Being that this file is placed in the `test` directory, Mocha will try to read and interpret it as a test suite, which it's not. Perhaps we could either inline these functions where they're used, or create a nested `fixtures` directory?

File: blocks/api/test/function-refs.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+ * Function references used so that Chai doesn't complain about functions
+ * not being equal.
+ *
+ * These are methods that should be assigned to a valid block.
+ */
+export const edit = function edit() {
```
Comment (aduth) [inline blocks/api/test/function-refs.js]: I think we could choose one or the other of the named variable or function:

```js
export function edit() {
```

```js
export const edit = function() {
// OR: 
//  export const edit = () => 'tacos';
```

The one-liner arrow function could work well as an inline function in each test, even if duplicated. Else I tend to prefer named functions with the function keyword.

File: blocks/api/test/parser.js [hunk 6]
```diff
@@ -99,15 +106,21 @@ describe( 'block parser', () => {
 		} );
 
 		it( 'should create the requested block with no attributes if it exists', () => {
-			registerBlock( 'core/test-block', {} );
+			registerBlock( 'core/test-block', {
+				edit: edit,
```
Comment (aduth) [inline blocks/api/test/parser.js]: Minor: We could take advantage of [ES2015 object property shorthand](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015) to simplify this:

```js
registerBlock( 'core/test-block', { edit, save } );
```

**Review by aduth — CHANGES_REQUESTED** [review]

File: blocks/api/registration.js (hunk 2 above)
Comment (BE-Webdesign) [inline blocks/api/registration.js]: > Should we inline block settings validation for consistency with the slug validation?

Most likely I am going to make another PR and totally redo the approach, to have the validation be all in one function rather than the sort of procedural thing we have developing.

> Should we expect the validate to do something, like throw, or perhaps prefixing with is to reflect its boolean return value is more accurate, like isValidBlockSettings

Prefixing sounds good.  We can throw as well.

> If we know the function returns a boolean, do we need to be so explicit with the comparison, or can we simplify to ! isValidBlockSettings ?

I am usually a fan of the strict checks, but can definitely change it.

File: blocks/api/test/function-refs.js (hunk 4 above)
Comment (BE-Webdesign) [inline blocks/api/test/function-refs.js]: Sure, I didn't see the harm in Mocha running it.  It pretty much just skips over it.

`fixtures` directory sounds good, I usually name it `test-data` is that good?

File: blocks/api/test/function-refs.js (hunk 5 above)
Comment (BE-Webdesign) [inline blocks/api/test/function-refs.js]: Yup, sounds good.

File: blocks/api/test/parser.js (hunk 6 above)
Comment (BE-Webdesign) [inline blocks/api/test/parser.js]: Yeah, I wasn't sure what style should be done for that.  In some projects people shy away from that as it can be confusing as to how that works.

File: blocks/api/registration.js (hunk 3 above)
Comment (BE-Webdesign) [inline blocks/api/registration.js]: Yup, I think the second one looks better to me.  Something I like to do in my apps is have a processing chain of sorts that has consistent input output behavior.  It works similar to `Promise`s.

This could just become a pipe ( flow in lowdash ) of `blockSettings` -> `isValid` -> `addBlock` -> return settings.  That way if done properly you can interchange the processing for whatever needs may come and in the future stick something new in there pretty painlessly.

**Comment by BE-Webdesign** [comment]
Rather than revising this right now, I think I may try a separate PR that demonstrates another way to handle this, if that approach seems more appealing, then we can go that way.  If we poo poo it, then we can come back to this one and I will revise.

**Comment by BE-Webdesign** [comment]
Closing, will circle back to this at another point in the future.

**Comment by swissspidy** [comment]
Can this branch be deleted after #1345?
